### ISSUE TITLE
HIP-1 Address Checksum Format

### ISSUE THREAD

**Issue by lbaird** [hashgraph/hedera-improvement-proposal#35]
This HIP addresses the request in issue #34. A [web page](https://codepen.io/lbaird/full/QWKRPGv) demonstrates this standard using embedded Javascript code. Example code can be downloaded for these languages:

- [AddressChecksums.java.zip](https://github.com/hashgraph/hedera-improvement-proposal/files/5861407/AddressChecksums.java.zip)
- [HIP-1_javascript.html.zip](https://github.com/hashgraph/hedera-improvement-proposal/files/5861376/HIP-1_javascript.html.zip)

### HIP-1 Address Checksum Format

Hedera defines two standard ways to write addresses: <i>with-checksum</i> or <i>no-checksum</i>. The with-checksum format consists of a no-checksum address followed by a 5-letter checksum. In the HTML JavaScript demonstration, enter an address in no-checksum format in the top box and click CONVERT to convert it to the with-checksum format in the second box. Or enter one manually in that box, and click VERIFY to see if it is valid or invalid.

### Format Definition

In the Hedera public ledger, each entity (account, file, smart contract, topic, token type) has an address such as `0.0.123` that is used to refer to it in transactions sent to the ledger.  In the past, it was typical for applications such as wallet software to use that format. However, a user attempting to type in such an address might accidentally leave out a digit, or swap two digits, or modify a digit. If the result is an actual account, this could result in hbars being transferred to the wrong account.

It is therefore useful to catch such errors before the transaction is sent to the network. So this HIP defines a standard for an alternative address format that is designed to help prevent such mistakes. The old no-checksum format continues to be valid, but the new with-checksum format extends it by appending a 5-letter checksum to the address. 

Software should be written to always display Hedera entity addresses in with-checksum format (such as `0.0.123-laujm`), with the checksum after a dash, all lowercase, and no spaces or other characters added. It should accept address inputs in either no-checksum ( `0.0.123` ) or with-checksum ( `0.0.123-laujm` ) format, all lowercase, with no additional whitespace or punctuation allowed, and no leading zeros for the integers. So these would both be accepted:

```
0.0.123
0.0.123-laujm
```

If the user enters any other format, or the checksum doesn't match, then the input should not be accepted, and the user should be told that it is incorrect, such as in these cases:

```
0.0.123-abcde
0.00.123
0.0.0123-laujm
0.0.123-LAUJM
0.0.123-lAuJm
0.0.123#laujm
0.0.123laujm
0.0.123 - laujm
0.123
0.0.123.
0.0.123-la
0.0.123-lau-jm
0.0.123-laujmxxxx
```

An address that is received as input should be rejected if it doesn't match the following regex. It should also be rejected if its checksum is incorrect.

`/^(0|(?:[1-9]\d*))\.(0|(?:[1-9]\d*))\.(0|(?:[1-9]\d*))(?:-([a-z]{5}))?$/`

An address that is displayed or sent as output should always be generated such that it would match the following regex. It is the same, except without the last `?`, so the checksum is always included.

`/^(0|(?:[1-9]\d*))\.(0|(?:[1-9]\d*))\.(0|(?:[1-9]\d*))(?:-([a-z]{5}))$/`

The checksum (such as `laujm`) is calculated from the no-checksum address (such as `0.0.123` ) by this algorithm:

```
a = a valid no-checksum address string, such as 0.0.123
d = int array for the digits of a (using 10 to represent "."), so 0.0.123 is [0,10,0,10,1,2,3]
h = unsigned byte array containing the ledger ID followed by 6 zero bytes
p3 = 26 * 26 * 26
p5 = 26 * 26 * 26 * 26 * 26
s0 = (d[0] + d[2] + d[4] + d[6] + ...) mod 11
s1 = (d[1] + d[3] + d[5] + d[7] + ...) mod 11
s = (...((((d[0] * 31) + d[1]) * 31) + d[2]) * 31 + ... ) * 31 + d[d.length-1]) mod p3
sh = (...(((h[0] * 31) + h[1]) * 31) + h[2]) * 31 + ... ) * 31 + h[h.length-1]) mod p5
c = (((d.length mod 5) * 11 + s0) * 11 + s1) * p3 + s + sh ) mod p5
checksum = c, written as 5 digits in base 26, using a-z
```
The checksum is a function of the ledger ID, so that the same address will have different checksums if it is on different ledgers. Cryptographically secure ledger IDs will be implemented as part of state proofs. But for now, the following three ledgers will each have a ledger ID consisting of a single byte:
```
0 = Hedera mainnet
1 = stable testnet
2 = preview net
```

The reference implementation is the Java code linked to at the top of this document. All implementations should match its outputs and behavior, including rejecting as invalid all the examples of invalid addresses above, and giving the the correct checksum in all of the following examples:

```
For ledger ID 0x00:
  0.0.1-auzeb
  0.0.123-laujm
  0.0.1234567890-ueafv
  12.345.6789-idmsv
  1.23.456-qzwsb

For ledger ID 0xa1ff01:
  0.0.1-ktdue
  0.0.123-uyyzp
  0.0.1234567890-ecevy
  12.345.6789-sbriy
  1.23.456-aybie
```

### Notes

The checksum is always 5 lowercase letters, and is guaranteed to catch any of the following errors:

- remove a digit (or 2 or 3 or 4 digits)
- add a digit (or 2 or 3 or 4 digits)
- modify a digit (or 2 adjacent digits)
- swap two different adjacent digits

Doing any of those modifications is guaranteed to change the checksum. In fact, it is guaranteed to change at least one of the first 2 letters of the checksum.  Furthermore, if the no-checksum part of the address were simply replaced with a random one, or the ledger ID were replaced with a random one, then it is extremely unlikely that the new address would have the same 5-character checksum as the old address (less than one in a million chance).

In the algorithm for calculating the checksum, the variable `s` is a weighted sum of all the digits (mod 26^3), `sh` is a weighted sum of all the bytes of the ledger ID padded with 6 zeros (mod 26^5), `s0` is a sum of the digits in the even positions (mod 11), and `s1` is a sum of the digits in the odd positions (mod 11).  If a digit is removed or added, then `(d.length mod 5)` will change.  If a digit in an even position is modified, then `s0` will change. If a digit in an odd position is modified, then `s1` will change. If two different adjacent digits are swapped, then both `s0` and `s1` will change. 

The 3 numbers `d.length, s0, s1` are encoded in the 2 most significant letters of the checksum, so if any of those conditions occur, the checksum will change in at least one of those two letters. 

The other 3 letters of the checksum are a very simple hash of the address. There are over 10 million different 5-letter checksums possible, so typos are likely to be caught, even if they aren't one of the 4 kinds of typos listed here.

Any function that creates 5-letter checksums will inevitably have collisions, where two addresses have the same checksum. If the checksum function were a cryptographically-strong pseudorandom function (PRF), then there would be collisions where the addresses differ in only a single digit. The function defined here has no collisions like that. 

Also, when calculating checksums for all accounts of the form `0.0.x` as `x` counts up 1, 2, 3, ..., a PRF would be expected to reach a collision within the first 3,500 numbers (around sqrt(26^5)), but the function here goes more than 10 times further. It reaches its first collision at `0.0.39004(vwmgo)`, which collides with the earlier address `0.0.10690(vwmgo)`. But that is fine, because it is unlikely that a person trying to type `0.0.39004(vwmgo)` would accidentally type `0.0.10690(vwmgo)`. And a more likely typo such as `0.0.3904` has a different checksum `(pgbgg)`, so an entered address of `0.0.3904(vwmgo)` would be flagged as incorrect by any application that follows this standard.

Last updated January 23, 2021.

The above and the example code are (c) 2020 Hedera Hashgraph, released under Apache 2.0 license.

**Comment by bugbytesinc** [comment]
While a bit of an edge case, have you considered including the network (mainnet, testnet, previewnet) as a part of the hash computation?

**Comment by lbaird** [comment]
That’s a good idea. The definition could be changed to include the ledger ID, once that has been implemented (as part of state proofs). Though it would mean we would have one set of hashes today, and then they would change when we implement ledger IDs. It’s worth thinking about.

**Comment by mehcode** [comment]
> While a bit of an edge case, have you considered including the network (mainnet, testnet, previewnet) as a part of the hash computation?

I was a bit on the fence of this being a good idea until I read this comment by @bugbytesinc. This is an excellent idea. Not only do we have some protection against mistyped IDs but we also have protection against someone sending to an ID of the wrong network which is an unsolved problem in Ethereum (when you copy address from a test wallet there are numerous warning pop-ups that try to help).

---

Can I also suggest that we follow precedence in the gaming industry for identifiers and use `0.0.10#check` as the format over `0.0.10(check)`. Its 1 less character to type and is used all over the place to disambiguate user identifiers (see Discord for a popular example).

**Comment by kenthejr** [comment]
@lbaird what do you think about @mehcode suggestion? If we can settle on format, I'd like to move this into a formal HIP review process.

**Comment by lbaird** [comment]
I think the 0.0.123(laujm) format is the least confusing possible for people who don't yet know we have a checksum. Because it's clear that the checksum is not part of the address, but is more like a comment on the address. Of course, any tool conforming with the above will still accept an inputof 0.0.123#laujm as valid. The parentheses are just the standard for how to display it.

At the moment, both Hedera and some outside parties are implementing the HIP as described above, with outputs using parentheses, and inputs accepting anything (parentheses or hash or anything else).

**Comment by mehcode** [comment]
I strongly disagree on that. Please specify that any other format but the one we're going with is considered illegal. 

Allowing clients to display and format how they want is an awful idea in practice. 

I'll concede on the format even though I'm pretty against how parentheses adds two symbols and the `#` adds one. 

---

Most importantly, can we please include the network in the hash? I find this to be as critical or even more so than protecting against typos.

**Comment by lbaird** [comment]
The standard above does NOT say that clients can display in any format they want. It says they should display with parentheses.  It also gives a regex that is lenient when parsing input. It follows the general principle of being strict for outputs, but lenient for inputs.

**Comment by lbaird** [comment]
Note that there are two regexes given: one is the requirement for outputs and the other is the requirement for inputs, to follow that principle.

**Comment by lbaird** [comment]
"Be conservative in what you send, be liberal in what you accept"

https://en.wikipedia.org/wiki/Robustness_principle

**Comment by kenthejr** [comment]
I'm reading two things we need to resolve before we settle on the HIP:

1. format for appending checksum
2. request to add network info

---

1. my recommendation would be to be highly opinionated in the format we recommend for input and output. The official SDKs will strictly implement those, but anyone can deviate (map their own format) however they want in the wild. I'm also a fan of fewer characters and a gamer, so `{accountID}#{checksum}` resonates with me. That would be my recommendation.
2. @mehcode I'm keen to see what you would recommend for network info.

**Comment by mehcode** [comment]
> "Be conservative in what you send, be liberal in what you accept"

Please see the criticism section of the wikipedia link you shared. Most internet standards in the last decade or so have increasingly not been following this for several very real reasons:

 - Flaws in implementations quickly become pseudo-standard and circulated as "well this is accepted here in client X?"

 - From a user experience perspective it is very helpful to give clear and concise errors to an input stream, being liberal in what is accepted makes this unreasonable in practice which leads to very confusing error messages.

 - From a performance perspective, being strict in what you accept leads to a significantly more performant implementation. Regular expressions require an entire embedded VM to utilize where as a straight-forward one-pass parse can handle a strict format.

 - Most notably this resonates in JSON and is a strong reason why JSON is used in API implementations as opposed to its brother YAML that follows the robustness principle. 

---

>  I'm keen to see what you would recommend for network info.

We have some kind of ID ( @lbaird mentioned a TBD "ledger ID" ) that is mixed in with the hash. My recommendation is to not wait on that to be defined and pick low constants to be mainnet and testnet. Other networks can always be different than those.

**Comment by bugbytesinc** [comment]
I like it, thanks for adding the network number, I've seen a few conversations in on discord where this feature would have prevented a human mistake.  

Is the expectation that the SDKs offer direct support in their APIs when building requests or is it expected implementation will solely be at the user interface level?  I gather this will not be included at the protobuf level initially?

**Comment by lbaird** [comment]
The expectation is that it would be used in SDKs, wallets, ledger explorers, etc. It wouldn’t be part of the HAPI protobuf transaction itself, but it would be part of the tools used to create and see the transactions.

________________________________

**Comment by rhysied** [comment]
Hi all, I had a few questions:

1) I know the likelihood of a collision bearing similar properties is very small, and that the lower the length of the checksum the more user friendly it is - however, is there some significance around having 5 as a length? Would 6/7/8 allow a much greater number of accounts before a collision is reached? Or would using 3/4 be more user friendly while achieving similar results?

2) do you think this has the potential to add spam to the network if people decide to create accounts with vanity checksums? Given the checksums can be calculated in advance I could know that, for example, by spending $20 I can create an account with some special significance?

Also, as regards to the rest of the discussion, in general I agree with Mehcode that more stringent input requirements/formatting seems better, however as for the format of ({checksum}) vs #{checksum}, I have no real preference.

Cheers, Rhys

**Comment by lbaird** [comment]
The 5 characters should be sufficient to catch typos. Because adding a character, deleting a character, and swapping two adjacent characters is guaranteed to never cause a collision.  Even adding or subtracting 2 or 3 or 4 characters is guaranteed to never cause a collision. So it seems unlikely that someone will accidentally mistype an address, and find that the resulting hash still matches.  In addition, if you make up two random addresses, the chance of collision is much less than one in a million.  So 5 characters should be sufficient for catching likely errors.

It could be reduced to 3 or 4 characters, and probably still be OK.  The number 5 was actually chosen because humans tend to remember 5 characters pretty easily, but 6 is harder. So it was the longest that is easy to remember. In the old days of cryptography, ciphertext was often written in groups of 5 digits for that reason.

It would be good to hear from others about whether to be strict on inputs or not. If it’s generally felt that strict is better, then I’ll edit the proposal to have just the second regex without the first one. Not all of the criticisms listed on the wiki page would apply in this case. But strict wouldn’t be terrible.

If there is widespread concern about wasting a byte, then we might consider 0.0.123-laujm instead of either 0.0.123(laujm) or 0.0.123#laujm because it is visually less crowded to use the lighter dash. I’d be fine with switching to that.

**Comment by rblinton** [comment]
prefer strict input and 0.0.123-laujm format

**Comment by lbaird** [comment]
From the various people who have commented, it sounds like it should be strict input with 0.0.123-laujm format. If there are no objections, I will change both the text and the reference implementation to reflect both of those changes.

**Comment by mehcode** [comment]
I'm slightly in favor of `0.0.123#laujm` over `0.0.123-laujm` because `#laujm` stands on its own and `-laujm` looks awkward. Let me try and explain with how this might look with pictures from MyHbarWallet (v1).

Now:

![image](https://user-images.githubusercontent.com/753919/105607709-3e137d00-5d55-11eb-9a50-8afa10f5c546.png)

After this spec:

![image](https://user-images.githubusercontent.com/753919/105607742-7c10a100-5d55-11eb-8d7a-17948306874d.png)

or

![image](https://user-images.githubusercontent.com/753919/105607747-85017280-5d55-11eb-89da-907967347a31.png)

---

I won't try to block this or anything on `#` vs `-`, just wanted to bring context for how it might look in a browser UI.

**Comment by bugbytesinc** [comment]
```
h = unsigned byte array containing the ledger ID followed by 6 zero bytes
```
It just occurred to me to ask, Is it important to know the network from the checksum just by looking at it?

I'm speculating not since it is trivial to compute a comparison value for all networks?

**Comment by rhysied** [comment]
I suppose the optimal thing for frontend applications to do would be to check any failures against the checksums for the account ID on different networks in case people are operating in the wrong context and throw a warning.

Not sure how you would enforce that though, it would just have to be a recommendation I guess, unless you wanted to go further down the rabbit hole you could potentially have a compliance program which allows sites / apps to show a "hedera standards compliant (vX.X)" badge
________________________________

**Comment by lbaird** [comment]
If a wallet application is designed to just work with mainnet, then it would be hard coded to use that ID for all checksums.

If it’s designed to work with several networks, such as the 3 that are listed, then it would presumably give the user a pull down menu to choose one of those three.

It could also have an option where the user can actually enter an arbitrary ledger ID.  In the future, we will be using cryptographically-secure ledger IDs of 48 bytes (a SHA-384 hash).  But for the moment we’re just using 7 bytes (a 0, 1, or 2 byte, followed by 6 zero bytes).

________________________________

**Comment by mehcode** [comment]
MyHbarWallet is going to add a network drop-down in the header and then it would require your checksum to  match the ID and the selected network.

**Comment by lbaird** [comment]
The HIP-1 description has now been updated in all 4 places:
- text description
- Javascript zip
- Java zip
- linked example page

**Comment by kenthejr** [comment]
Let's get this converted to the new HIP format. We can continue to use this issue to discuss the HIP once the PR is submitted and reviewed.
